### ISSUE TITLE
Tracking Issue for binary heap retain

### ISSUE THREAD

**Issue by arlopurcell** [rust-lang/rust#71503]
This is a tracking issue for `BinaryHeap::retain`.
The feature gate for the issue is `#![feature(binary_heap_retain)]`.

Added in PR #71485

**Comment by KodrAus** [comment]
We discussed this at the recent Libs meeting and had a question about taking `&T` rather than `&mut T`. In the end we thought keeping consistent with `Vec` and not having to think about resorting afterwards made the `&T` item type appropriate.

If anybody is keen to open a stabilization PR we can kick off a FCP!

**Comment by JohnTitor** [comment]
> If anybody is keen to open a stabilization PR we can kick off a FCP!

@rust-lang/libs-api May I ask to kick off a FCP? Stabilizing it will unblock rust-lang/rust-clippy#9059.

**Comment by Amanieu** [comment]
@rfcbot fcp merge

**Comment by Amanieu** [comment]
I think we should change the API to either:
- Use `&mut` for the closure argument (which is consistent with `HashMap`)
- Or add `retain_mut` like `Vec` does.

I personally prefer the first option: `retain_mut ` was only added to `Vec` because `retain` was already stabilized when we noticed the mistake.

@rfcbot concern &mut

**Comment by JohnTitor** [comment]
> I think we should change the API to either:
> 
>     * Use `&mut` for the closure argument (which is consistent with `HashMap`)
> 
>     * Or add `retain_mut` like `Vec` does.
> 
> 
> I personally prefer the first option: `retain_mut ` was only added to `Vec` because `retain` was already stabilized when we noticed the mistake.

Sounds good, should it be addressed in a stabilization PR or done before it?

**Comment by Amanieu** [comment]
This can be addressed in the stabilization PR but we do need to make a decision here.

**Comment by m-ou-se** [comment]
> I personally prefer the first option: `retain_mut ` was only added to `Vec` because `retain` was already stabilized when we noticed the mistake.

Agreed.

**Comment by joshtriplett** [comment]
Agreed: let's make it `retain`, and perhaps mention something in its documentation for people who might be looking for `retain_mut`.

**Comment by Amanieu** [comment]
It seems we have a consensus to make `retain` take a closure with `&mut T`.

@rfcbot resolve &mut

**Comment by ijackson** [comment]
I don't think this is right.

As I understand it, a `retain` whose `F` takes `&T` can be implemented more efficiently since it can know that the ordering doesn't change.  For example, if only a small number of elements are removed, a few log-time fixups would suffice.

A `retain` which gives out `&mut T` needs to assume the ordering might change, and would have to check it and be prepared to reinsert elements somewhere entirely else.

**Comment by ijackson** [comment]
(This is unlike `Vec` which doesn't care about what the contents are like, so `Vec::retain_mut` is just the same as `Vec::retain` inside.)

**Comment by dtolnay** [comment]
I agree with @ijackson's rebuttal; `retain` and `retain_mut` are fundamentally different operations for BinaryHeap. #78681 is already taking advantage of the distinction. I don't think I would want to add the behavior of retain_mut under the name `retain`.

**Comment by dtolnay** [comment]
@rfcbot concern predicate panic can leave heap in invalid state

#105851 introduced an invariant that BinaryHeap methods are required to abide by. The current implementation of `retain` does not meet it.

```rust
#![feature(binary_heap_retain)]

use std::collections::BinaryHeap;
use std::panic::{self, AssertUnwindSafe};

fn main() {
    let mut heap = BinaryHeap::from_iter([3, 1, 2]);
    println!("heap={:?}", heap);
    panic::catch_unwind(AssertUnwindSafe(|| {
        heap.retain(|e| {
            if *e == 1 {
                panic!();
            }
            false
        });
    }));
    println!("heap={:?}", heap);
    println!("sorted={:?}", heap.into_sorted_vec());
}
```

This prints `sorted=[2, 1]` which is obviously not a sorted vector.

**Comment by dtolnay** [comment]
I'm marking yaahc's box because she has stepped down from T-libs-api after the point that this feature got proposed for FCP.

This can enter FCP as soon as the predicate panic concern is addressed.

**Comment by Aaron1011** [comment]
It looks like https://github.com/rust-lang/rust/pull/106918 fixed the predicate panic issue - can this now enter FCP?

**Comment by dtolnay** [comment]
@rfcbot resolve predicate panic can leave heap in invalid state
I fixed this in #106918.
